File: projects/wnba-edge-lab/pipeline/ledger.py
"""Model shadow-book grading from final ESPN scores.

The published site never auto-adds a model pick to the user's ledger. Actual
wagers live only in the browser ledger after the user clicks Add. This module
keeps the separate shadow book needed to audit every model tier honestly.
"""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timezone


def _result(row: dict, game: dict) -> str | None:
    if not game.get("completed"):
        return None
    away_score = game["away"].get("score")
    home_score = game["home"].get("score")
    if away_score is None or home_score is None:
        return None
    away_score, home_score = float(away_score), float(home_score)
    market, side = row["market"], row["side"]
    if market == "ML":
        mine, theirs = (away_score, home_score) if side == "away" else (home_score, away_score)
        return "Win" if mine > theirs else "Loss" if mine < theirs else "Push"
    if market == "ATS":
        mine, theirs = (away_score, home_score) if side == "away" else (home_score, away_score)
        adjusted = mine + float(row.get("line") or 0.0)
        return "Win" if adjusted > theirs else "Loss" if adjusted < theirs else "Push"
    if market == "TOTAL":
        total = away_score + home_score
        line = float(row.get("line") or 0.0)
        if total == line:
            return "Push"
        return "Win" if (side == "over") == (total > line) else "Loss"
    return None


def _profit(stake: float, price: float, result: str) -> float:
    if result in {"Push", "Void"}:
        return 0.0
    if result == "Loss":
        return -stake
    return stake * (price / 100.0 if price > 0 else 100.0 / -price)


def _lock(row: dict, tracked_at: str, shadow: bool = False) -> dict:
    return {
        "id": row["candidate_id"],
        "game_id": row["game_id"],
        "date": row["date"],
        "tipoff": row["tipoff"],
        "matchup": row["matchup"],
        "market": row["market"],
        "side": row["side"],
        "pick": row["pick"],
        "line": row.get("line"),
        "price": row["price"],
        "book": row.get("book"),
        "model_prob": row["model_prob"],
        "breakeven": row["breakeven"],
        "edge": row["edge"],
        "confidence": row["confidence"],
        "tier": row["tier"],
        "stake": 1.0 if shadow else row["stake"],
        "tracked_at": tracked_at,
        "result": "Pending",
        "profit": 0.0,
        "shadow": shadow,
    }


def sync(candidates: list[dict], games: list[dict], ledger_state: dict | None,
         shadow_state: dict | None, now: str | None = None) -> tuple[dict, dict]:
    now = now or datetime.now(timezone.utc).isoformat(timespec="seconds")
    game_map = {game["game_id"]: game for game in games}
    # Previous releases wrote qualifying model picks here automatically. Clear
    # that generated list during the first refresh after this migration. The
    # actual wager ledger is now explicit browser-local state, like MLB Edge.
    ledger = {"bets": [], "mode": "manual-browser"}
    shadow = shadow_state or {"calls": []}
    shadow.setdefault("calls", [])

    # The shadow book records every priced side so tier quality can be tested.
    shadow_existing = {row["id"] for row in shadow["calls"]}
    for row in candidates:
        from .model_accuracy import instant
        start = instant(row.get("tipoff"))
        if not start or start <= instant(now) or (game_map.get(row["game_id"]) or {}).get("completed"):
            continue
        if row["candidate_id"] not in shadow_existing:
            shadow["calls"].append(_lock(row, now, shadow=True))
            shadow_existing.add(row["candidate_id"])

    for row in shadow["calls"]:
        if row.get("result") not in {None, "Pending"}:
            continue
        game = game_map.get(row["game_id"])
        if not game:
            continue
        result = _result(row, game)
        if result:
            row["result"] = result
            row["profit"] = round(_profit(float(row["stake"]), float(row["price"]), result), 2)
            row["graded_at"] = now
            row["final_score"] = f'{int(game["away"]["score"])}-{int(game["home"]["score"])}'
    ledger["updated_at"] = now
    shadow["updated_at"] = now
    return ledger, shadow


def performance(ledger: dict, shadow: dict, starting_bankroll: float) -> dict:
    bets = ledger.get("bets") or []
    settled = [row for row in bets if row.get("result") != "Pending"]
    profit = round(sum(float(row.get("profit") or 0.0) for row in settled), 2)
    risked = round(sum(float(row.get("stake") or 0.0) for row in settled), 2)

    def groups(rows: list[dict], key: str) -> dict:
        buckets: dict[str, list[dict]] = defaultdict(list)
        for row in rows:
            buckets[str(row.get(key) or "Unknown")].append(row)
        output = {}
        for name, bucket in buckets.items():
            graded = [row for row in bucket if row.get("result") != "Pending"]
            wins = sum(row.get("result") == "Win" for row in graded)
            losses = sum(row.get("result") == "Loss" for row in graded)
            pushes = sum(row.get("result") == "Push" for row in graded)
            wagered = sum(float(row.get("stake") or 0.0) for row in graded)
            pnl = sum(float(row.get("profit") or 0.0) for row in graded)
            output[name] = {
                "record": f"{wins}-{losses}-{pushes}",
                "settled": len(graded),
                "win_pct": round(wins / (wins + losses), 4) if wins + losses else None,
                "profit": round(pnl, 2),
                "roi": round(pnl / wagered, 4) if wagered else None,
            }
        return output

    calls = shadow.get("calls") or []
    return {
        "starting_bankroll": starting_bankroll,
        "current_bankroll": round(starting_bankroll + profit, 2),
        "pending_bets": sum(row.get("result") == "Pending" for row in bets),
        "settled_bets": len(settled),
        "profit": profit,
        "risked": risked,
        "roi": round(profit / risked, 4) if risked else None,
        "by_tier": groups(calls, "tier"),
        "by_market": groups(calls, "market"),
    }

File: projects/wnba-edge-lab/site/ledger.js
/* My Ledger — only wagers the user explicitly confirms. */
(function (root, factory) {
  const api = factory();
  if (typeof module === "object" && module.exports) module.exports = api;
  else root.WNBALedger = api;
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  const STORAGE_KEY = "wnbaedge.ledger.v2";
  const SCHEMA = 2;

  const decimal = american => {
    const price = Number(american);
    if (!isFinite(price) || price === 0) return 1;
    return 1 + (price > 0 ? price / 100 : 100 / Math.abs(price));
  };

  const keyOf = row => `${row.game_id}|${row.market}|${row.side}`;

  function entryFrom(row, stakeOverride) {
    const stake = stakeOverride == null ? Number(row.stake || 0) : Number(stakeOverride);
    return {
      id: keyOf(row), candidate_id: row.candidate_id, game_id: row.game_id,
      date: row.date, tipoff: row.tipoff, matchup: row.matchup,
      market: row.market, side: row.side, pick: row.pick,
      line: row.line == null ? null : Number(row.line), price: Number(row.price),
      book: row.book || null, model_prob: Number(row.model_prob),
      market_fair_prob: row.market_fair_prob == null ? null : Number(row.market_fair_prob),
      breakeven: Number(row.breakeven), edge: Number(row.edge), tier: row.tier,
      stake: Math.round(Math.max(0, stake) * 100) / 100,
      added_at: new Date().toISOString(), result: null, profit: null, final_score: null,
    };
  }

  function settle(entry, game) {
    if (!game || !game.completed || !game.away || !game.home ||
        game.away.score == null || game.home.score == null) return null;
    const away = Number(game.away.score), home = Number(game.home.score);
    let result = null;
    if (entry.market === "ML") {
      const mine = entry.side === "away" ? away : home;
      const theirs = entry.side === "away" ? home : away;
      result = mine > theirs ? "Win" : mine < theirs ? "Loss" : "Push";
    } else if (entry.market === "ATS") {
      const mine = entry.side === "away" ? away : home;
      const theirs = entry.side === "away" ? home : away;
      const adjusted = mine + Number(entry.line || 0);
      result = adjusted > theirs ? "Win" : adjusted < theirs ? "Loss" : "Push";
    } else if (entry.market === "TOTAL") {
      const total = away + home, line = Number(entry.line);
      if (!isFinite(line)) return null;
      result = total === line ? "Push" : ((total > line) === (entry.side === "over") ? "Win" : "Loss");
    }
    if (!result) return null;
    const stake = Number(entry.stake || 0);
    const profit = result === "Win" ? stake * (decimal(entry.price) - 1)
                 : result === "Loss" ? -stake : 0;
    return {result, profit: Math.round(profit * 100) / 100,
            final_score: `${away}-${home}`, settled_at: new Date().toISOString()};
  }

  function settleAll(entries, games) {
    const byId = new Map((games || []).map(game => [String(game.game_id), game]));
    let changed = 0;
    const rows = entries.map(entry => {
      if (entry.result) return entry;
      const result = settle(entry, byId.get(String(entry.game_id)));
      if (!result) return entry;
      changed++;
      return Object.assign({}, entry, result);
    });
    return {entries: rows, changed};
  }

  function summarise(entries, startingBankroll) {
    const settled = entries.filter(row => row.result);
    const profit = settled.reduce((sum, row) => sum + Number(row.profit || 0), 0);
    const risked = settled.filter(row => row.result !== "Push")
      .reduce((sum, row) => sum + Number(row.stake || 0), 0);
    const wins = settled.filter(row => row.result === "Win").length;
    const losses = settled.filter(row => row.result === "Loss").length;
    const pushes = settled.filter(row => row.result === "Push").length;
    return {
      bankroll: Math.round((Number(startingBankroll || 0) + profit) * 100) / 100,
      profit: Math.round(profit * 100) / 100, risked: Math.round(risked * 100) / 100,
      roi: risked ? profit / risked : null, wins, losses, pushes,
      pending: entries.filter(row => !row.result).length, settled: settled.length,
      at_risk: Math.round(entries.filter(row => !row.result)
        .reduce((sum, row) => sum + Number(row.stake || 0), 0) * 100) / 100,
    };
  }

  function load() {
    try {
      const parsed = JSON.parse(localStorage.getItem(STORAGE_KEY) || "null");
      const rows = Array.isArray(parsed) ? parsed : (parsed && parsed.entries) || [];
      return rows.filter(row => row && row.game_id && row.market && row.side);
    } catch (_) { return []; }
  }

  function save(entries) {
    try {
      localStorage.setItem(STORAGE_KEY, JSON.stringify({schema: SCHEMA,
        saved_at: new Date().toISOString(), entries}));
      return true;
    } catch (_) { return false; }
  }

  function merge(existing, incoming) {
    const byKey = new Map(existing.map(row => [keyOf(row), row]));
    let added = 0;
    incoming.forEach(row => {
      if (!row || !row.game_id || !row.market || !row.side) return;
      const key = keyOf(row);
      if (!byKey.has(key)) { byKey.set(key, row); added++; }
    });
    return {entries: [...byKey.values()], added};
  }

  const CSV_COLS = ["date","matchup","market","side","pick","line","price","book",
    "stake","tier","edge","model_prob","result","profit","final_score","added_at"];
  function toCSV(entries) {
    const quote = value => {
      const text = value == null ? "" : String(value);
      return /[",\n]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
    };
    return [CSV_COLS.join(",")].concat(entries.map(row => CSV_COLS.map(key => quote(row[key])).join(","))).join("\n");
  }

  return {STORAGE_KEY, SCHEMA, decimal, keyOf, entryFrom, settle, settleAll,
          summarise, load, save, merge, toCSV, CSV_COLS};
});
